### ISSUE TITLE
CRUD abstraction and implementation.

### ISSUE THREAD

**Issue by okedoki** [nreco/data#4]
I have implemented the CRUD, but I still have a few issues. The issues are described in the second message.

First, about the attached files. 
Idea is to create a layer depends only on Dictionaries, modify/comment/use them. That realization isn't included the Batch operations, but will be included in the next version.

[СRUD.zip](https://github.com/nreco/data/files/418044/RUD.zip)

The basic idea is to create an abstraction level, for that operation you have to put a list of the following class

```
        public class CRUDInfo
        {
            /// <summary>
            /// Table name
            /// </summary>
            public string TableName;
            /// <summary>
            /// ID name
            /// </summary>
            public string IdName;


            public KeyValuePair<string, object> IdPair
            {
                get
                {
                    return new KeyValuePair<string, object>(IdName, ObjectDictionary[IdName]);
                }
            }
            /// <summary>
            /// Dictionary contains dictionary with datatable name of columns and values 
            /// </summary>


            public IDictionary<string, object> ObjectDictionary;
}
```

CRUD operations:

```
            abstract internal void Create(IEnumerable<CRUDInfo> data);
            abstract internal IEnumerable<CRUDInfo> Retrieve(IEnumerable<CRUDInfo> rInfo);
            abstract internal void Update(IEnumerable<CRUDInfo> data);
            abstract internal void Delete(IEnumerable<CRUDInfo> data);
```

For example, you can pass a following list to Create

```
   crudInfoList.Add( new CRUDInfo()
            {
                TableName = "People",
                IdName = "PeopleId",
                ObjectDictionary = new Dictionary<string, object>()
            {
                { "PeopleId", "123"},
                { "Name", "John"},
            }
            };);
    crudInfoList.Add( new CRUDInfo()
            {
                TableName = "Pets",
                IdName = "Id",
                ObjectDictionary = new Dictionary<string, object>()
            {
                { "Id", "1"},
                { "Name", "John' pet"},
                { "PeopleFK", "123"},
            }
            };);
///This command should create the two records
Create(crudInfoList);
```

**Comment by okedoki** [comment]
Two issues what I have at this moment:

1) I didn't find a way to use create a command to work with several rows for CRUD.
For example, I expect to select records with ID =1,2,3. How can I do it?
Now, I have to do it in the loop.
```
foreach (CRUDInfo retriveInfo in rInfoList)
                        {
                            q.Condition = ((QField)retriveInfo.IdPair.Key == new QConst(retriveInfo.IdPair.Value));
                            selectComand = dbContext.CommandBuilder.GetSelectCommand(q);
                            dbContext.InitCommand(selectComand);

                            using (var rdr = selectComand.ExecuteReader())
                            {
                                while (rdr.Read())
                                {
                                    retrievedList.Add(new CRUDInfo()
                                    {
                                        TableName = retriveInfo.TableName,
                                        ObjectDictionary = Enumerable.Range(0, rdr.FieldCount).ToDictionary(rdr.GetName, rdr.GetValue)
                                    });
                              
  }
                            }
                        } 
```
2) If use update or create, I can't use IDictionary only Dictionary. Can it be modified?
 CRUDInfo.ObjectDictionary is an IDictionary<string, object>,but I can't pass it directly to GetUpdateCommand, only through the wrapper.

 ```
var changeset = new Dictionary<string, object>(CRUDInfo.ObjectDictionary);
 updateCommand = dbContext.CommandBuilder.GetUpdateCommand(q, changeset) 
```

**Comment by VitaliyMF** [comment]
@okedoki  regarding CRUDInfo, it seems you're trying to implement the concept that looks similar to old ADO.NET DataTable/DataRow - could you check issue https://github.com/nreco/data/issues/2 (RecordSet implementation)? I've already added first draft of this structure: [RecordSet.cs](https://github.com/nreco/data/blob/master/src/NReco.Data/RecordSet.cs). Conceptually RecordSet works like DataTable and suitable for processing many records at once. Right now I'm working on DbDataAdapter.Select().ToRecordSet() and DbDataAdapter.Update(RecordSet) implementations -- I assume that these enhancements may make your CRUDInfo structure unnecessary. 

> 1) I didn't find a way to use create a command to work with several rows for CRUD.
> For example, I expect to select records with ID =1,2,3. How can I do it?

why not make one select for selecting several records at once?.. like that:

```
var recordIds = new int[] { 1, 2, 3}; 
new Query(someTableName, new QConditionNode( (QField)"id", Conditions.In, new QConst(recordIds) ) );
```

> 2) If use update or create, I can't use IDictionary only Dictionary. Can it be modified?

Oh I see, I forgot to add overload for IDictionary<string,object> in DbCommandBuilderExtensions. Give me 10 minutes :-)

**Comment by VitaliyMF** [comment]
@okedoki take a look to the updated [SqliteDemo.DataAdapter](https://github.com/nreco/data/blob/master/examples/SqliteDemo.DataAdapter/Program.cs) -- I've added sample code for dictionary-driven CRUD operations, and new RecordSet.

Also all DbCommandGenerator / DbDataAdapter methods support IDictionary<string,object>. I've just published 1.0.0-alpha4 and you can try it.

**Comment by VitaliyMF** [comment]
It seems this question is no longer actual for latest 1.0-final release.
